---
title: "PBMC_stimulated"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# This dataset is taken from the following study:

# Kang, H., Subramaniam, M., Targ, S. et al. Multiplexed droplet single-cell RNA-sequencing using natural genetic variation. Nat Biotechnol 36, 89â€“94 (2018). https://doi.org/10.1038/nbt.4042



```

```{r}
library(Seurat)
library(ggplot2)
library(edgeR)
library(patchwork)
library(scmap)

devtools::load_all(path="../scDECAF")

GetMSigDB <- function(database="Hallmarks"){
  
  
  Hallmark_Geneset_DIR <- "GeneSets/h.all.v7.2.symbols.gmt"
  C2_DIR <- "GeneSets/c2.all.v7.2.symbols.gmt"
  # if (substr(getwd(), 2, 9) == "stornext"){
  #   Hallmark_Geneset_DIR="/stornext/Home/data/allstaff/w/whitfield.h/BIG_DATA/GeneSets/h.all.v7.0.symbols.gmt"
  #   C2_DIR="/stornext/Home/data/allstaff/w/whitfield.h/BIG_DATA/GeneSets/c2.all.v7.1.symbols.gmt"
  # } else {
  #   Hallmark_Geneset_DIR="C:\\Users\\whitfield.h\\Desktop\\Projects\\B_I_G_Data\\GeneSets\\MSigDB\\msigdb.v7.2.symbols.gmt"
  #   C2_DIR="C:\\Users\\whitfield.h\\Desktop\\Projects\\B_I_G_Data\\GeneSets\\MSigDB\\c2.all.v7.1.symbols.gmt"
  # }

  if (database == "Hallmarks") {
    Hallmark_Genesets <- list()
    conn <- file(Hallmark_Geneset_DIR,open="r")
    linn <-readLines(conn)
    for (i in 1:length(linn)){
      iLine <- linn[i]
      iLine <- strsplit(iLine, "\t")
      iName <- iLine[[1]][1]
      iList <- iLine[[1]]
      iList <- iList[3:length(iList)]
      Hallmark_Genesets[[iName]] <- iList
     # Hallmark_Genesets <- c(Hallmark_Genesets, list(GeneSet(setName=iName, geneIds=iList)))
    }
    close(conn)
    return(Hallmark_Genesets)
    
  } else if (database == "C2") {
    C2_Genesets <- list()
    conn <- file(C2_DIR,open="r")
    linn <-readLines(conn)
    for (i in 1:length(linn)){
      iLine <- linn[i]
      iLine <- strsplit(iLine, "\t")
      iName <- iLine[[1]][1]
      iList <- iLine[[1]]
      iList <- iList[3:length(iList)]
      C2_Genesets[[iName]] <- iList
      #C2_Genesets <- c(C2_Genesets, list(GeneSet(setName=iName, geneIds=iList)))
    }
    close(conn)
    return(C2_Genesets)
  } else if (database == "Both"){
    Genesets <- list()
    conn <- file(Hallmark_Geneset_DIR,open="r")
    linn <-readLines(conn)
    for (i in 1:length(linn)){
      iLine <- linn[i]
      iLine <- strsplit(iLine, "\t")
      iName <- iLine[[1]][1]
      iList <- iLine[[1]]
      iList <- iList[3:length(iList)]
      Genesets[[iName]] <- iList

    }
    close(conn)

    conn <- file(C2_DIR,open="r")
    linn <-readLines(conn)
    for (i in 1:length(linn)){
      iLine <- linn[i]
      iLine <- strsplit(iLine, "\t")
      iName <- iLine[[1]][1]
      iList <- iLine[[1]]
      iList <- iList[3:length(iList)]
      Genesets[[iName]] <- iList

    }
    close(conn)
    
    return(Genesets)
  }

}


HM_genesets <- GetMSigDB(database="Hallmarks")
C2_genesets <- GetMSigDB(database="C2")
MSigDB <- GetMSigDB(database="Both")


xcontrol <- readRDS("PBMC_interferon/PBMC_control.RDS")
xstim <- readRDS("PBMC_interferon/PBMC_interferon-stimulated.RDS")

pbmc <-  merge(CreateSeuratObject(xcontrol, project = "control"),
                  y = CreateSeuratObject(xstim, project = "interferon-stimulated"),
                  add.cell.ids = c("control", "interferon-stimulated"),
                  project = "PBMC-stim")

pbmc


pbmc <- SCTransform(object = pbmc, verbose = FALSE, variable.features.n = 1000)
pbmc



pbmc <- RunUMAP(object = pbmc, features = VariableFeatures(object = pbmc),
                  reduction.key = 'sctUMAP_', reduction.name = 'sctumap')


Idents(pbmc) <- pbmc$orig.ident
p1 <- DimPlot(pbmc, cols = "Paired") + theme_void()

```

```{r}
dim(x = GetAssayData(object = pbmc, slot = 'data'))
# x <- GetAssayData(object = pbmc, slot = 'scale.data') 
x <- GetAssayData(object = pbmc, slot = 'data')# same as x <- bcdata[["SCT"]]@data, contains log-normalised values
x[1:10,1:10]


hvg <- VariableFeatures(object = pbmc)

# target <- genesets2ids(x[match(hvg, rownames(x)),], HM_genesets)
target <- genesets2ids(x, HM_genesets)
colSums(target)


target <- target[, colSums(target) > 0]


cell_embeddings <- pbmc@reductions$sctumap@cell.embeddings
cell_embeddings <- cell_embeddings[match(colnames(x), rownames(cell_embeddings)),]

# not required as the function internally takes care after this
# x <- as.matrix(x)

ann_res <- scDECAF(data = x, gs = target, standardize = TRUE, 
                   hvg = hvg, k = 5, embedding = cell_embeddings,
                   n_components = ncol(target) - 1, max_iter = 2, thresh = 0.5)


#head(attributes(ann_res)$raw_scores)

interferon_alpha_scores <- attributes(ann_res)$raw_scores[,"HALLMARK_INTERFERON_ALPHA_RESPONSE"]
summary(interferon_alpha_scores)


pbmc <- AddMetaData(pbmc, scale(interferon_alpha_scores), "HALLMARK_INTERFERON_ALPHA_RESPONSE")
p2 <- FeaturePlot(pbmc, features = "HALLMARK_INTERFERON_ALPHA_RESPONSE", cols = c( "white", "red"),
                  #cells = names(bcdata$pred_celltype[grepl("Macrophage", bcdata$pred_celltype)]),
                 pt.size = 0.05, order = TRUE) + theme_void()



interferon_gamma_scores <- attributes(ann_res)$raw_scores[,"HALLMARK_INTERFERON_GAMMA_RESPONSE"]
summary(interferon_gamma_scores)


pbmc <- AddMetaData(pbmc, scale(interferon_gamma_scores), "HALLMARK_INTERFERON_GAMMA_RESPONSE")
p3 <- FeaturePlot(pbmc, features = "HALLMARK_INTERFERON_GAMMA_RESPONSE", cols = c( "white", "red"),
                  #cells = names(bcdata$pred_celltype[grepl("Macrophage", bcdata$pred_celltype)]),
                 pt.size = 0.05, order = TRUE) + theme_void()



cowplot::plot_grid(p1,p2,p3, ncol=2)


ggdat <- data.frame(scores = c(interferon_alpha_scores, interferon_gamma_scores),
                    hallmark = rep(c("IFN-alpha","IFN-gamma"), times = c(length(interferon_alpha_scores), length(interferon_gamma_scores))),
                    phenotype = rep(pbmc$orig.ident,2))


ggplot(ggdat, aes(x=hallmark, y = scores, fill = phenotype)) +
  geom_boxplot()


# Dimplot but on cells predicted as IFN related

pbmc$hallmark_preds <- ann_res$pred_celltype
Idents(pbmc) <- "hallmark_preds"

p4 <- DimPlot(pbmc, cells = ann_res$cell[ann_res$pred_celltype %in% c("HALLMARK_INTERFERON_GAMMA_RESPONSE",
                                                                            "HALLMARK_INTERFERON_ALPHA_RESPONSE",
                                                                      "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
                                                                      "HALLMARK_INFLAMMATORY_RESPONSE",
                                                                      "HALLMARK_KRAS_SIGNALING_DN",
                                                                      "HALLMARK_KRAS_SIGNALING_UP",
                                                                      "HALLMARK_MYC_TARGETS_V1")], 
              cols = "Paired", pt.size = 0.1) + theme_void()




cowplot::plot_grid(p1,p4,p2,p3, ncol=2)

table(pbmc$orig.ident, pbmc$hallmark_preds)

```


```{r}
summary(attributes(ann_res)$raw_scores)
```



```{r}

selected_gs <- pruneGenesets(data = x, genesetlist = HM_genesets, hvg = hvg,
                            embedding = cell_embeddings, min_gs_size = 3, lambda = exp(-2))


as.character(selected_gs)
```



```{r}
target <- genesets2ids(x[match(hvg, rownames(x)),], HM_genesets[selected_gs])

dim(x)
dim(target)



ann_res <- scDECAF(data = x, gs = target, standardize = TRUE, 
                   hvg = hvg, k = 5, embedding = cell_embeddings,
                   n_components = ncol(target) - 1, max_iter = 2, thresh = 0.5)



summary(attributes(ann_res)$raw_scores)


interferon_alpha_scores <- attributes(ann_res)$raw_scores[,"HALLMARK_INTERFERON_ALPHA_RESPONSE"]
summary(interferon_alpha_scores)


pbmc <- AddMetaData(pbmc, scale(interferon_alpha_scores), "HALLMARK_INTERFERON_ALPHA_RESPONSE")
p2 <- FeaturePlot(pbmc, features = "HALLMARK_INTERFERON_ALPHA_RESPONSE", cols = c("#FDE725FF", "#51127CFF"),
                  #cells = names(bcdata$pred_celltype[grepl("Macrophage", bcdata$pred_celltype)]),
                 pt.size = 0.05, order = TRUE) + theme_void() + theme(legend.position = "None",
                                                                      plot.background = element_rect(colour = "black", size = 0.5))




ggdat <- data.frame(scores = c(interferon_alpha_scores),
                    hallmark = rep(c("IFN-alpha"), times = c(length(interferon_alpha_scores))),
                    phenotype = rep(pbmc$orig.ident,1))


ggplot(ggdat, aes(x=hallmark, y = scores, fill = phenotype)) +
  geom_boxplot()


myc <- attributes(ann_res)$raw_scores[,"HALLMARK_MYC_TARGETS_V1"]
pbmc <- AddMetaData(pbmc, scale(myc), "HALLMARK_MYC_TARGETS_V1")
pz <- FeaturePlot(pbmc, features = "HALLMARK_MYC_TARGETS_V1", cols = c("#FDE725FF", "#51127CFF"),
                  #cells = names(bcdata$pred_celltype[grepl("Macrophage", bcdata$pred_celltype)]),
                 pt.size = 0.05, order = TRUE) + theme_void() + theme(legend.position = "None",
                                                                      plot.background = element_rect(colour = "black", size = 0.5))




p1 + pz



homeo <- attributes(ann_res)$raw_scores[,"HALLMARK_CHOLESTEROL_HOMEOSTASIS"]
pbmc <- AddMetaData(pbmc, scale(homeo), "HALLMARK_CHOLESTEROL_HOMEOSTASIS")
px <- FeaturePlot(pbmc, features = "HALLMARK_CHOLESTEROL_HOMEOSTASIS", cols = c("#FDE725FF", "#51127CFF"),
                  #cells = names(bcdata$pred_celltype[grepl("Macrophage", bcdata$pred_celltype)]),
                 pt.size = 0.05, order = TRUE) + theme_void() + theme(legend.position = "None",
                                                                      plot.background = element_rect(colour = "black", size = 0.5))




secre <- attributes(ann_res)$raw_scores[,"HALLMARK_PROTEIN_SECRETION"]
pbmc <- AddMetaData(pbmc, scale(secre), "HALLMARK_PROTEIN_SECRETION")
py <- FeaturePlot(pbmc, features = "HALLMARK_PROTEIN_SECRETION", cols = c("#FDE725FF", "#51127CFF"),
                  #cells = names(bcdata$pred_celltype[grepl("Macrophage", bcdata$pred_celltype)]),
                 pt.size = 0.05, order = TRUE) + theme_void() + theme(legend.position = "None",
                                                                      plot.background = element_rect(colour = "black", size = 0.5))




oxy <- attributes(ann_res)$raw_scores[,"HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY"]
pbmc <- AddMetaData(pbmc, scale(oxy), "HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY")
ps <- FeaturePlot(pbmc, features = "HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY", cols = c("#FDE725FF", "#51127CFF"),
                  #cells = names(bcdata$pred_celltype[grepl("Macrophage", bcdata$pred_celltype)]),
                 pt.size = 0.05, order = TRUE) + theme_void() + theme(legend.position = "None",
                                                                      plot.background = element_rect(colour = "black", size = 0.5))


angio <- attributes(ann_res)$raw_scores[,"HALLMARK_ANGIOGENESIS"]
pbmc <- AddMetaData(pbmc, scale(angio), "HALLMARK_ANGIOGENESIS")
pt <- FeaturePlot(pbmc, features = "HALLMARK_ANGIOGENESIS", cols = c("#FDE725FF", "#51127CFF"),
                  #cells = names(bcdata$pred_celltype[grepl("Macrophage", bcdata$pred_celltype)]),
                 pt.size = 0.05, order = TRUE) +
   theme_void() + theme(#legend.position = "None",
                        plot.background = element_rect(colour = "black", size = 0.5))




cowplot::plot_grid(p1 + theme(plot.background = element_rect(colour = "black", size = 0.4)),p2, pz, px, #py, #ps,
                   pt,
                   nrow = 1)
```

